### PR TITLE
Simplify connection closing and cleanups in TcpTransport

### PR DESCRIPTION
**Pull request by s1monw**
Today we maintain a map of open connections in order to close them when
a low level channel gets closed or handles a failure. We also spawn a thread due to some
tricky concurrency issues especially with respect to netty since they listener might
be called on a transport / boss thread. Executions on those threads must not be blocking
since otherwise we will likely deadlock the event processing which adds to the
complexity of the concurrency model in this class.

This change associates the connection with the close callback that every channel invokes
once it's closed which allows us to remove the connections map. A relaxed non-blocking
concurrency model in the connection close listener allows cleaning up connected nodes without
blocking on any lock.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 1]
```diff
@@ -465,6 +459,20 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
                         logger.debug("connected to node [{}]", node);
                     }
                     transportServiceAdapter.onNodeConnected(node);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: paranoia: if we hit an exception here, we can leave the nodeChannels in the connectedNodes map. Should we move this:

```
  if (connectedNodes.remove(node, nodeChannels)) {   
      transportServiceAdapter.onNodeDisconnected(node);
  }
```

to the finally clause (under the check of success)?

**Review by bleskes — APPROVED** [review]
This is great. Thanks for pursuing it further

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 2]
```diff
@@ -953,7 +940,7 @@ protected void onException(Channel channel, Exception e) {
         String reason = ExceptionsHelper.detailedMessage(e);
```
Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: Unused?

File: core/src/main/java/org/elasticsearch/transport/TcpTransport.java [hunk 3]
```diff
@@ -557,77 +572,38 @@ public final NodeChannels openConnection(DiscoveryNode node, ConnectionProfile c
...
-                    return true;
-                }
-            }
-        }
-        return false;
-    }
-
-    private void closeAndNotify(DiscoveryNode node, NodeChannels nodeChannels, String reason) {
+    private void disconnectFromNodeCloseAndNotify(DiscoveryNode node, NodeChannels nodeChannels) {
+        assert nodeChannels != null : "nodeChannels must note be null";
```
Comment (tbrooks8) [inline core/src/main/java/org/elasticsearch/transport/TcpTransport.java]: Nit: spelling

**Review by Tim-Brooks — APPROVED** [review]
Just a few small things

**Pull request merged**